addons/keda/keda.go

```diff
@@ -0,0 +1,551 @@
+	camelv1 "github.com/apache/camel-k/pkg/apis/camel/v1"
+	camelv1alpha1 "github.com/apache/camel-k/pkg/apis/camel/v1alpha1"
+	"github.com/apache/camel-k/pkg/kamelet/repository"
+	"github.com/apache/camel-k/pkg/metadata"
+	"github.com/apache/camel-k/pkg/platform"
+	"github.com/apache/camel-k/pkg/trait"
+	"github.com/apache/camel-k/pkg/util"
+	"github.com/apache/camel-k/pkg/util/kubernetes"
+	"github.com/apache/camel-k/pkg/util/property"
+	"github.com/apache/camel-k/pkg/util/source"
+	"github.com/apache/camel-k/pkg/util/uri"
+	"github.com/pkg/errors"
+	scase "github.com/stoewer/go-strcase"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	// kameletURNMetadataPrefix allows binding Kamelet properties to KEDA metadata.
+	kameletURNMetadataPrefix = "urn:keda:metadata:"
+	// kameletURNAuthenticationPrefix allows binding Kamelet properties to KEDA authentication options.
+	kameletURNAuthenticationPrefix = "urn:keda:authentication:"
+	// kameletURNRequiredTag is used to mark properties required by KEDA.
+	kameletURNRequiredTag = "urn:keda:required"
+
+	// kameletAnnotationType indicates the scaler type associated to a Kamelet.
+	kameletAnnotationType = "camel.apache.org/keda.type"
+	// kameletAnnotationMetadataPrefix is used to define virtual metadata fields computed from Kamelet properties.
+	kameletAnnotationMetadataPrefix = "camel.apache.org/keda.metadata."
+	// kameletAnnotationAuthenticationPrefix is used to define virtual authentication fields computed from Kamelet properties.
+	kameletAnnotationAuthenticationPrefix = "camel.apache.org/keda.authentication."
+)
+
+// The KEDA trait can be used for automatic integration with KEDA autoscalers.
+// The trait can be either manually configured using the `triggers` option or automatically configured
+// via markers in the Kamelets.
+//
+// For information on how to use KEDA enabled Kamelets with the KEDA trait, refer to
+// xref:ROOT:kamelets/kamelets-user.adoc#kamelet-keda-user[the KEDA section in the Kamelets user guide].
+// If you want to create Kamelets that contain KEDA metadata, refer to
+// xref:ROOT:kamelets/kamelets-dev.adoc#kamelet-keda-dev[the KEDA section in the Kamelets development guide].
+//
+// The KEDA trait is disabled by default.
+//
+// +camel-k:trait=keda.
+type kedaTrait struct {
+	trait.BaseTrait `property:",squash"`
+	// Enables automatic configuration of the trait. Allows the trait to infer KEDA triggers from the Kamelets.
+	Auto *bool `property:"auto" json:"auto,omitempty"`
+	// Convert metadata properties to camelCase (needed because Camel K trait properties use kebab-case from command line). Disabled by default.
+	CamelCaseConversion *bool `property:"camel-case-conversion" json:"camelCaseConversion,omitempty"`
+	// Set the spec->replicas field on the top level controller to an explicit value if missing, to allow KEDA to recognize it as a scalable resource.
```

Comment:
I need to investigate.. yeah I called it "hack" because it smells :)

Reply:
Turns out that the error is thrown by Kube: error 500 when KEDA tries to get the scale subresource: https://github.com/kedacore/keda/blob/9ac7a0a28e446335376ef8f5e82494cf05456ecb/controllers/keda/scaledobject_controller.go#L293.

I've already seen this in the past...

Reply:
Would it be possible that the Integration hasn't been reconciled yet and the `.status.replicas` field hasn't been set when the Keda controller reconciles the ScaledObject? If that is the case, maybe we could have it set by default earlier, either when the Integration is created, or even possibly defaulted in the CRD?

Reply:
No, the `.status.replicas` field is not considered. Here's a test I ran with a fully deployed klb (with Keda trait disabled).

```
$ kubectl get klb sqs-to-telegram -o "jsonpath={.status.replicas}"
1

$ curl --silent http://localhost:8088/apis/camel.apache.org/v1alpha1/namespaces/test/kameletbindings/sqs-to-telegram/scale | jq
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {},
  "status": "Failure",
  "message": "Internal error occurred: the spec replicas field \".spec.replicas\" does not exist",
  "reason": "InternalError",
  "details": {
    "causes": [
      {
        "message": "the spec replicas field \".spec.replicas\" does not exist"
      }
    ]
  },
  "code": 500
}
```

So the error comes from the API server.
Then, if I set the `.spec.replicas` everything goes well:

```
$ kubectl get klb sqs-to-telegram -o "jsonpath={.spec.replicas}"


$ kubectl scale klb sqs-to-telegram --replicas 1
kameletbinding.camel.apache.org/sqs-to-telegram scaled


$ kubectl get klb sqs-to-telegram -o "jsonpath={.spec.replicas}"
1

$ curl --silent http://localhost:8088/apis/camel.apache.org/v1alpha1/namespaces/test/kameletbindings/sqs-to-telegram/scale | jq
{
  "kind": "Scale",
  "apiVersion": "autoscaling/v1",
  "metadata": {
    "name": "sqs-to-telegram",
    "namespace": "test",
    "uid": "0b93916c-72f1-47ae-b2d1-74f9e3f7e5b6",
    "resourceVersion": "512151",
    "creationTimestamp": "2022-01-03T22:56:18Z"
  },
  "spec": {
    "replicas": 1
  },
  "status": {
    "replicas": 1,
    "selector": "camel.apache.org/integration=sqs-to-telegram"
  }
}
```

May it be my version of Kube... it's not latest, but it's a recent one:

```
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"22", GitVersion:"v1.22.2", GitCommit:"8b5a19147530eaac9476b0ab82980b4088bbc1b2", GitTreeState:"clean", BuildDate:"2021-09-15T21:38:50Z", GoVersion:"go1.16.8", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"22", GitVersion:"v1.22.1", GitCommit:"632ed300f2c34f6d6d15ca4cef3d3c7073412212", GitTreeState:"clean", BuildDate:"2021-08-19T15:39:34Z", GoVersion:"go1.16.7", Compiler:"gc", Platform:"linux/amd64"}
```

Reply:
Right. I wonder what's the best practice in that case. Ultimately I can see two options:
- Setting the replicas spec field is optional, then Keda logic that checks the scale sub-resource should be improved, to only fail on 404, at least handling the 500 return code gracefully
-  Otherwise, it's mandatory, then it may be better to set it as default value in the CRD directly

We could choose the later option anyway, as it makes the default value explicit. WDYT?

Reply:
+1 to make the value explicit

Reply:
I'm realising setting it in the CRD may not generalise well for Knative, as it'll deactivate autoscaling. Also it may not really be relevant for CronJob in its current form.

Thinking about the Knative case, I would find it a bit confusing to have a replicas spec set, but an actual replicas status to never eventually converge towards the spec, given the autoscaling. That actually applies to Keda as well. It seems they make the assumption that the replicas spec must be set, for the sole purpose of checking the existence of the `scale` sub-resource.

As a user, I'd be inclined to expect that, when I set the replicas spec, the replicas status eventually converges towards the spec. If that's just me 😃, we can remove the logic that forces that behaviour for Knative, so that it's consistent across auto-scalers. I also wonder how HPA handles this case...

Reply:
Thanks a lot for the feedback. I'm not sure I read this correctly, but my interpretation of that documentation is that `specReplicasPath` is mandatory in the CustomResourceDefinition, but the field specified by the `specReplicasPath` in the custom resource is not, in which case the `/scale` subresource returns an error on GET.

Similarly, `statusReplicasPath` is mandatory, but the field that's pointed in the custom resource is not, in which case the status replica value in the `/scale` subresource defaults to 0.

That indicates fields are mandatory in the CRD, but the corresponding fields in the custom resources are not. If that interpretation is correct, an error on GET'ing the `/scale` sub resource does not necessary implies the absence of the `/scale` sub resource, it can also implies the field defined by `specReplicasPath` in the custom resource is not set.

Reply:
Yeah, that's my understanding as well.

Reply:
If we follow that reasoning, would the ideal solution be that Keda amends it's _scale_ sub-resource detection logic, to better handle the case where GET'ing the `/scale` endpoint errors because the field defined by `specReplicasPath` has no value, and not because the resource does not have the _scale_ sub-resource defined in its CRD?

Reply:
How is `specReplicasPath` and `statusReplicasPath` defined in the CRD? In KEDA we are using standra go-client call to check the `scale` subresource, so I'd like to avoid some hacks that might cause problems with some other resources.
Bear in mind that `/scale` subresource is being queried by HPA to drive the scaling, so if it is not correctly defined, then it won't work.
If you think that the CRD is correct and problem is in `client-go` we should try to fix it over there.

Reply:
> That indicates fields are mandatory in the CRD, but the corresponding fields in the custom resources are not. If that interpretation is correct, an error on GET'ing the /scale sub resource does not necessary imply the absence of the /scale sub resource, it can also imply the field defined by specReplicasPath in the custom resource is not set.

I missed this paragraph before. I mean that those fields need to point to an actual fields that actually returns and sets a correct value. So the HPA controller (and KEDA as well) can correctly drive the scaling.

Reply:
They are defined with:
https://github.com/apache/camel-k/blob/e9553940554ea533aaae2d0074681f37b434e7aa/config/crd/bases/camel.apache.org_integrations.yaml#L5631-L5635

My understanding is that the detection logic in `checkTargetResourceIsScalable` could be updated to handle the 500 code case with message "the spec replicas field \".spec.replicas\" does not exist", because the resource is scalable in that case:

https://github.com/kedacore/keda/blob/9ac7a0a28e446335376ef8f5e82494cf05456ecb/controllers/keda/scaledobject_controller.go#L293

Reply:
`.spec.replicas ` and `.status.replicas ` exists and are correctly set as well?

Reply:
Let's use our Integration example. The `specReplicasPath` is set in the CRD:

https://github.com/apache/camel-k/blob/e9553940554ea533aaae2d0074681f37b434e7aa/config/crd/bases/camel.apache.org_integrations.yaml#L5631-L5635

But the `.spec.replicas` in the Integration resource is not set by default.

According to the documentation:

> If there is no value under the specReplicasPath in the custom resource, the /scale subresource will return an error on GET.

The following command fails:

```console
$ kubectl  --raw /apis/camel.apache.org/v1/namespaces/test/integrations/my-integration/scale | jq
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {},
  "status": "Failure",
  "message": "Internal error occurred: the spec replicas field \".spec.replicas\" does not exist",
  "reason": "InternalError",
  "details": {
    "causes": [
      {
        "message": "the spec replicas field \".spec.replicas\" does not exist"
      }
    ]
  },
  "code": 500
}
```

Still the resource has a _scale_ sub-resource and is scalable. It can be scaled, for example with `kubectl scale`.

Reply:
Right, here is what I'm observing:
- It is possible to define a CRD with a scale sub-resource, and have instances of that CRD, i.e. custom resources of that CRD, created without values for the corresponding `specReplicasPath` CRD field, e.g. `.spec.replicas`. The API server allows it, and the documentation indicates this is allowed.
- In that case a GET request to the resource `/scale` endpoint returns 500 with the message `the spec replicas field \".spec.replicas\" does not exist"
- On the other hand, for a custom resource whose CRD does not define a scale sub-resource, a GET request to the resource `/scale` endpoint returns 404.

In both cases, I find the result sensible:
- 500 means, the sub-resource exists, and the custom resource is scalable, but the server just cannot determine the current scale value, because it is not set
- 404 means, the `/scale` resource is not found, the resource is not scalable.

Reply:
Yeah, that might be true. But how do we get the current replicas then, if quering `/scale` results in 500 (ie. not set replica count)?

Later in the code, KEDA queries this endpoint to find out what is the current state and whether it should modify the replica count. This wouldn't work then.

https://github.com/kedacore/keda/blob/53a81e9f12e9c1833b66e2e0e6fe514db1e9e47e/pkg/scaling/executor/scale_scaledobjects.go#L64
->
https://github.com/kedacore/keda/blob/53a81e9f12e9c1833b66e2e0e6fe514db1e9e47e/pkg/scaling/executor/scale_scaledobjects.go#L255

Reply:
Yeah this might be seen as obvious approach, but unfortunately it doesn't work this way. There are certain states where we need to compare the actual number of replicas to multiple values in ScaledObject and act based on the result. It is not just assignig one value as a replica count.

I mean, everything could be done, right? :) So maybe if we refactor to code a lot it could be done, but I am not 100% sure about this. It would definitely make the code less logical.

Reply:
Okay, I thought about it more and it cannot be done, there are certain KEDA features that won't work. For example there's setting to reset replica count back to original number, once ScaledObject is deleted. If we don't know the number of replicas in the moment when we create the ScaledObject, we don't know the value that we should use once ScaledObject is deleted. 
And there definitely other things, that would be broken.

Reply:
Right, sorry to drag you on this. I'm totally OK with KEDA mandating the `.spec.replicas` field to be set. I'm still trying to understand what's the best decision on this in the general case.

Reply:
> If we don't know the number of replicas in the moment when we create the ScaledObject, we don't know the value that should be used to restore the replicas count, once ScaledObject is deleted.
For that case, it was not set, so it could be unset. It may be a dubious parallel, but it's like a pointer in Golang, when it's needed to differentiate between the default value for the type, and not initialised.

Reply:
> Right, sorry to drag you on this. I'm totally OK with KEDA mandating the `.spec.replicas` field to be set. I'm still trying to understand what's the best decision on this in the general case.

Yeah, no worries :) It is always great to discuss stuff, it gives you another perspective.

> For that case, it was not set, so it could be unset. It may be a dubious parallel, but it's like a pointer in Golang, when it's needed to differentiate between the default value for the type, and not initialised.

But behind those numbers (pointers), there are actuall replicas of a workload that needs to be set. I am not sure what unset means in this case :)